### ISSUE TITLE
Get scalar field at spatial position, instead of directly B field

### ISSUE THREAD

**Issue by Sylk1-9** [magpylib/magpylib#607]
### Describe the feature you'd like

I would find it very interesting and useful for many applications to be able to get the scalar potential at a given spatial point, similar to the `getB` routine.

Similarly, a routine to get the magnetic field gradient would also have many interesting applications. Although one can compute it numerically directly from the B field.

Hats off for your work.

**Comment by OrtnerMichael** [comment]
Hey @Sylk1-9 

Thanks for pointing this out.
Unfortunately the scalar potential has a gauge degree of freedom which means that, if evaluated in a single position, any value is correct. The scalar potential is only well-defined in relation to values at other positions, which makes it a bit ambiguous to work with within the framework of Magpylib.

Also there are problems, like e.g. the CylinderSegment, for which a B-field solution exists, but no scalar potential solution.

Also, there is the problem that the scalar potential can only be used when there are no free currents - and we have implemented a bunch of those - so you could only compute it in special cases.

Finally, I'm not really sure what is the use of the scalar potential, except to compute the B-field.

So long story short, there are a lot of reasons not implement it - also it's a lot of work. But we are open for discussion. If you have an idea how to overcome these issues !

We agree that the gradient fields would be very nice to have :), but as you pointed out, it is easy to compute with finite differences ... so we never made the extra effort to search for and implement these fields. Also - its a hell of an effort because you have three times as many gradient fields as you have fields.

Whats your motivation to have the scalar potential and the gradient fields ?

**Comment by Sylk1-9** [comment]
Hum, that is a lot of good reasons ! Thanks for taking the time to develop them. I didn't even know about the non-existing solution for the CylinderSegment for example.

My motivation is initially to interpolate the magnetic field on a mesh. Unfortunately, using usual polynomial (cubic for example) does not guarantee that Maxwell equations (grad B = 0 and div B = 0) are fulfilled. There exists literature about sets of polynomials (called harmonic polynomial) that does fulfil those two equations.
Here is two interesting papers,
    - [High-Performance Interpolation of Stellarator Magnetic Fields - Paul Probert](https://hsx.wisc.edu/documents/high-performance-interpolation-of-stellarator-magnetic-fields/) : here, a mathematical framework for interpolating either the B field or the scalar field is derived.
    - [Processing magnetic field data - H. Wind](https://www.sciencedirect.com/science/article/pii/0021999168900570) : This one interpolate both the B magnetic field and its gradient.

Another motivation is that less interpolation coefficients are needed, as the set of interpolation functions for a given power order is somehow more restricted. In the second reference, the number of interpolation coefficients to compute is reduced from 105 (4th degree polynomial) to only 35.

I noticed that magpylib does provide tools to [interpolate linearly](https://magpylib.readthedocs.io/en/stable/examples/examples_22_field_interpolation.html).
Extending the library to include those harmonic polynomial technics could be a nice upgrade as well.
I am anyway currently starting working on this subject, and could come back later with some code and examples that could be integrated to magpylib if you are interested.

Best regards.

**Comment by OrtnerMichael** [comment]
That looks interesting but please consider the following:
1. The analytical solutions provided by Magpylib are so fast that they will be only slightly slower than the evaluation of a polynomial. It might not be worth it to work on an interpolation scheme.
2. Magpylib does not offer Interpolation. In the docu we have an example where we still do it using Scipy, but the purpose of the example was to demonstrate the Custom class
3. We would be interested in advanced interpolation techniques though - in case you have some insights on this topic :)

**Comment by Sylk1-9** [comment]
1)  To give even more background around my motivations: I am currently working on simulation of particle trajectories in a magnetic field. The programme is written in c++. 

As a first approached, I successfully embedded magpylib into it. Unfortunately, the simulation is slow for, I guess, two main reasons : 
- It cannot use the vectorization offered by magpylib, as the programme is a time-step integrator, and by construction, doesn't know in advance for which positions the magnetic field has to be computed. 
- The number of source is quite high (something like, 17000 current loops distributed spatially, and 800 CylinderSegment).

Hence, my interest toward the use of a precomputed mesh + polynomial interpolation. 

3) I'll come back later with some propositions for grid interpolation, if you are interested then ;)

**Comment by OrtnerMichael** [comment]
Understood - this is a good motivation !!!

Specifically the CylinderSegment is a slow thing :( - The equations are pretty complex
